Rename reserved function names

pfl opened the pull request:
Rename reserved function names in drivers/ subdirectory. Update
function macros concatenatenating function names with '##'. As
there is a conflict between the existing gpio_sch_manage_callback()
and _gpio_sch_manage_callback() names, leave the latter unmodified.

While waiting for the prefix discussion to conclude, add also renaming for arch/.

The reserved function names are discussed in #9882

pfl committed:
drivers: Rename reserved function names

Rename reserved function names in drivers/ subdirectory. Update
function macros concatenatenating function names with '##'. As
there is a conflict between the existing gpio_sch_manage_callback()
and _gpio_sch_manage_callback() names, leave the latter unmodified.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

mike-scott commented on drivers/modem/wncm14a2a.c:
```diff
@@ -265,7 +265,7 @@ static inline void _hexdump(const u8_t *packet, size_t length)
 }
 #else
-#define _hexdump(...)
+#define z_hexdump(...)
```
Why is this z_hexdump when the renaming was from _hexdump() -> hexdump()?

ceolin replied:
yeah, symbol is part of this driver so does not need have zephyr's namespace IMHO

mike-scott replied:
I meant `z_hexdump` doesn't exist.  `_hexdump` was renamed to `hexdump` (minus the underscore)

ceolin replied:
I see, my comment is still valid though

pfl replied:
Renaming accident, I'll fix it up to just hexdump().

jhedberg commented on include/kernel.h:
```diff
@@ -1387,7 +1387,7 @@ struct k_timer {
 };
 
-#define _K_TIMER_INITIALIZER(obj, expiry, stop) \
+#define Z_TIMER_INITIALIZER(obj, expiry, stop) \
```
This seems very unintuitive to me. This API operates on an object called k_timer, so whatever the prefix is (or isn't) the name should still contain K_TIMER

andrewboie replied:
I disagree.
These are private APIs, this is why the name had a leading underscore. End users shouldn't be using this. The naming convention for private objects is to prefix with z_ instead of k_

tbursztyka replied:
What about Z_K_TIMER ? Though it's a bit ugly, it does keep the full object type name. Knowing also that 'K' has a strong meaning, unlike Z... unless we rename all the kernel types k_<> to z_<> but that would be even weirder.

jhedberg replied:
I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? If that's the case, how do you propose getting rid of users like this: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L72

andrewboie replied:
> I'm not opposed to the z_, just that the name of the API should contain the object that it operates on (it's a k_timer not a z_timer).

I don't really have a strong feeling about this one way or another, aesthetically z_k_ seems awkward but it's just a few macros. I'll go with the prevailing opinion.

> Btw, what exactly do you mean by "internal"? That only the k_timer code should be using this? 

I am not sure where our policy is documented on subsys/ with regards to private kernel APIs.

For those macros in particular, IIRC this was due to objects declared with them not being accessible to user mode.

The solution I'm currently pursuing is to implement futexes, they don't any restrictions where their memory can live.

jhedberg replied:
Btw, either of these changes will likely increase the chance of application code accessing the internal APIs, since for an uneducated/naive developer who's looking at include/kernel.h a z_* API from a project called Zephyr will not look in anyway suspicious or communicate a "stay away!" message. I realize the ship may already have sailed, but I'd have at least considered putting any internal APIs into a separate `kernel_internal.h`, and if we want to be completely clear to the user throw an `_internal` suffix at them.

pfl replied:
The leading underscore is the one targeted, any other prefix not starting with it is fine once we agree on it.

ceolin replied:
keep prefix `k` for private kernel API is ok for me. @pfl what do you think ? I can't disagree with @jhedberg points.

pfl replied:
Fine with me, but you'd need to spell out if you want a subset of the former '_k_*' functions or all of them changed to prefix 'k_'. I'm going to assume all, so I'll update the patch set accordingly.

The difficult part of this discussion is that it is nowhere explicitly stated which functions are actually exported to the rest of the system ("kernel" and "user space"), and which are "kernel" internal. Johan's comment about include/ and kernel/include is spot on, even I wouldn't know the difference.

We probably need to document this and start moving include files into the same hierarchy we now have for ext/, subsys/ etc.. Right now include/ is a bit too messy when just trying to comprehend available features by reading the individual .h files and trying to figure out if they are ext/, subsys/ or something elsel. Going in the direction from Kconfig variables to finding the relevant include files is the way it seems to work now.

andrewboie replied:
> keep prefix k for private kernel API is ok for me

No, please do not do this.
These private APIs, which can disappear or change at *any* time, need to be clearly distinguishable from Zephyr's public interfaces!
z_ makes this very clear that this is a different namespace.

jhedberg replied:
> These private APIs, which can disappear or change at any time, need to be clearly distinguishable from Zephyr's public interfaces!

@andrewboie I completely agree with this.

> z_ makes this very clear that this is a different namespace.

Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naïve reader the private APIs will not look private at all.

pfl replied:
> What makes things worse is that currently these are in the same public header file as the public APIs and (at least some) private APIs don't even have a code comment saying they're private. For a naive reader the private APIs will not look private at all.

But that is then a bit of a different issue. I'm not happy at all that the private/public/subsys/ext header files are mixed into the same directory, at minimum the public and private APIs should be in different files and the files properly ordered into a hierarchy. That seems to be a task for 1.15, though.

jhedberg replied:
@pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

pfalcon replied:
> Indeed, but not that it's a private API. What makes things worse is that currently these are in the same public header file as the public APIs

+1 for splitting public into private APIs into separate headers. That's the best way to keep namespace clean. Except that in our case it will resolve the situation only "so much", as e.g. we have private macros used to define public APIs. Still, it's the best mechanism we'd have to structure and organize Zephyr namespace.

pfl replied:
> @pfl understood, and agreed mostly. I just think that (ignoring the standard compliance, reserved namespace issue) the _ prefix had at least a slightly better chance of screaming "private!" than z_ has.

I agree here. But as we want to be MISRA C complieant, a prefix of '_' is out of bounds anyway.

jhedberg commented on include/kernel.h:
```diff
@@ -2049,12 +2049,12 @@ struct k_fifo {
  * @cond INTERNAL_HIDDEN
  */
-#define _K_FIFO_INITIALIZER(obj) \
+#define Z_FIFO_INITIALIZER(obj) \
```
Same here. The object operated on is called k_fifo, so the name should contain K_FIFO

jhedberg commented on include/kernel.h:
```diff
@@ -3006,7 +3006,7 @@ struct k_sem {
 };
 
-#define _K_SEM_INITIALIZER(obj, initial_count, count_limit) \
+#define Z_SEM_INITIALIZER(obj, initial_count, count_limit) \
```
Same here.

pfl committed:
arch: Rename reserved function names

Rename reserved function names in arch/ subdirectory. The Python
script gen_priv_stacks.py was updated to follow the 'z_' prefix
naming.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl commented:
Added arch/ function renaming, as it anyway is based on the changes in drivers/.

aescolar commented on arch/posix/include/posix_arch_internal.h:
```diff
@@ -9,13 +9,13 @@
 #include "toolchain.h"
 
-#define _SAFE_CALL(a) _safe_call(a, #a)
+#define Z_SAFE_CALL(a) z_safe_call(a, #a)
```
Please rename this PC_SAFE_CALL and pc_safe_call respectively (this macro and function are meant to be private to the posix arch, and should not appear to be usable by anybody else)

pfl replied:
Fixed.

pfl committed:
boards: Rename reserved function names

Rename reserved function names in the boards/ subdirectory.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
soc: Rename reserved function names

Rename reserved function names in the soc/ subdirectory.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl committed:
subsystems: Rename reserved function names

Rename reserved function names in the subsys/ subdirectory except
for static _mod_pub_set and _mod_unbind functions in bluetooth mesh
cfg_srv.c which clash with the similarly named global functions.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

pfl commented:
Fixed up reserved function names in boards, soc and subsys. The functions @jhe commented about are not yet renamed; these can in the worst case be renamed in a separate new patch in this PR.

vanti commented on ext/hal/ti/simplelink/source/ti/devices/cc32xx/inc/hw_gpio.h:
```diff
@@ -465,7 +465,7 @@
 #define GPIO_O_GPIO_PCTL        0x0000052C  // This register is not used in
                                             // cc3xx. equivalant register exsist
                                             // outside GPIO IP (refer
-                                            // PAD*_config register in the
```
I don't think this file needs to be changed. The comments are referring to some register name, not Zephyr function name.

pfl replied:
Automatic name conversion bug, fixed.

vanti replied:
LGTM

pfl commented:
> 1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "subsys: Rename reserved function names"

Hrm, this commit follows the above pattern *exactly*...

ioannisg reviewed (commented):
OK for soc/arm/nordic_nrf changes

Laczen reviewed (approved):
ok for subsys/fs/nvs

vanwinkeljan reviewed (approved):
ok for display related changes

pfl commented:
Updated to latest master.

pfl commented:
Conflicts resolved, updated to latest master.

andrewboie commented:
All,

We really need to make progress on this. The real issue seems to be that there is some controversy on whether the Z_ prefix is really appropriate to namespace private APIs, some feel that Z_ does not have the "don't touch this" vibe of the old _ prefix scheme.

So I think the real discussion is, "what should the new private API prefix be"?

We can discuss here or in #9882.

But to kick things off, how about "pvt_" instead of "z_" for this?

aescolar commented:
> So I think the real discussion is, "what should the new private API prefix be"?

Whatever it would be, before this PR is merged, it needs to be clearly decided what is the convention for all these. Renaming APIs is an annoyance for anybody who is implementing anything which uses Zephyr, so we should not change them back and forth. If we need to, let's please do it only once.

andrewboie commented:
> Renaming APIs is an annoyance for anybody who is implementing anything which uses Zephyr

What's under discussion is how to name *private* APIs. These already are deemed subject to renaming, changed semantics, or outright removal at any time.

ceolin commented:
> All,
> 
> We really need to make progress on this. If we punt to 1.15, any backporting of patches to 1.14 is going to be an ordeal.
> 
> The real issue seems to be that there is some controversy on whether the Z_ prefix is really appropriate to namespace private APIs, some feel that Z_ does not have the "don't touch this" vibe of the old _ prefix scheme.
> 
> So I think the real discussion is, "what should the new private API prefix be"?
> 
> We can discuss here or in #9882.
> 
> But to kick things off, how about "pvt_" instead of "z_" for this?

The problem @jhedberg's mentioned  is that with have objects prefixed with `k_` (e.g `k_timer`). So, is it ok we use `k_` instead of using `z_` ? or rename `k_` objects to `z_` ?

I completely agree about having it before 1.14 btw.

andyross commented:
Grumpy curmudgeon here:  **JUST PICK SOMETHING, THEY'RE PRIVATE APIS**.

They aren't documented.  They aren't legal to use.  No one has complained about z_ over the past year or so, which is about when it started appearing in the tree.

If there is a better choice for our private APIs to use, then we can go in and rename them all in 1.15, which will be perfectly fine because they're private APIs we can change whenever we like.  Then, in 1.16, when the "we ARE the system"[1] rule exception makes it through all the relevant international regulatory agencies, we can go back and change it to an underscore for clarity and concision.  And that will be perfectly fine too.

[1] No seriously: the reason underscores are disallowed is that they've been used historicaly in toolchain and system headers for internal symbols.   SunOS/AIX/Solaris/BSD/Linux headers don't follow this rule.  Ours needn't either, and for the same reason.  But whatever.  Just pick something.

andrewboie commented:
> The problem @jhedberg's mentioned is that with have objects prefixed with k_ (e.g k_timer). So, is it ok we use k_ instead of using z_ ? or rename k_ objects to z_ ?

Definitely not.

How are you going to distinguish private vs public APIs if you name everything the same?

> If there is a better choice for our private APIs to use, then we can go in and rename them all in 1.15, which will be perfectly fine because they're private APIs we can change whenever we like. Then, in 1.16, when the "we ARE the system"[1] rule exception makes it through all the relevant international regulatory agencies, we can go back and change it to an underscore for clarity and concision. And that will be perfectly fine too.

Amen! Half the kernel is already using the z_ convention...if nobody has a credible alternative in the new few days just roll with it for 1.14 IMO

pfl commented:
> Amen! Half the kernel is already using the z_ convention...if nobody has a credible alternative in the new few days just roll with it for 1.14 IMO

Time to hit the merge button then, before new conflicts arise and I have to rebase the set again.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -49,7 +49,7 @@ static void zsock_received_cb(struct net_context *ctx,
 			      void *user_data);
 
-static inline int _k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
+static inline int k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
```
The naming of this function is not that great atm. This private function looks now like a global function. Imho we need to change the name to something like `wait_non_empty()` or `fifo_wait_non_empty()`. This could be done in this PR or I can send a separate PR after this one is merged.

jukkar commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -26,7 +26,7 @@ extern const struct socket_op_vtable sock_fd_op_vtable;
 static const struct socket_op_vtable can_sock_fd_op_vtable;
 
-static inline int _k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
+static inline int k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
```
same here

jukkar commented on subsys/net/lib/sockets/sockets_packet.c:
```diff
@@ -26,7 +26,7 @@ extern const struct socket_op_vtable sock_fd_op_vtable;
 static const struct socket_op_vtable packet_sock_fd_op_vtable;
 
-static inline int _k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
+static inline int k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
```
and here too. I wonder if we should place it into a library, either into global one or just local to sockets.

ioannisg commented on arch/arm/core/swap_helper.S:
```diff
@@ -22,7 +22,7 @@ _ASM_FILE_PROLOGUE
 GTEXT(__pendsv)
-GTEXT(_do_kernel_oops)
+GTEXT(z_do_kernel_oops)
 GTEXT(_arm_do_syscall)
```
Why _arm_do_syscall is not updated?

pfl replied:
It's one of those cases that exist only in .S files. The starting point for these changes was the Coccinelle script, and Coccinelle only detects code in .h and .c files. Doing a manual fixup for this.

ioannisg reviewed (commented):
Looks ok. Left a comment about _arm_do_syscalls

pfl commented:
Fixed ioannis' issue and rebased.
